src/awkward/_v2/operations/structure/ak_concatenate.py

```diff
@@ -4,524 +4,201 @@
+
+                length = max(
+                    [len(x) for x in inputs if isinstance(x, ak._v2.contents.Content)]
+                )
+                nextinputs = []
+                for x in inputs:
+                    if isinstance(x, ak._v2.contents.Content):
+                        nextinputs.append(x)
+                    else:
+                        nextinputs.append(
+                            ak._v2.contents.ListOffsetArray(
+                                ak._v2.index.Index64(
+                                    nplike.arange(length + 1, dtype=np.int64)
+                                ),
+                                ak._v2.contents.NumpyArray(
+                                    nplike.broadcast_to(nplike.array([x]), (length,))
+                                ),
+                            )
+                        )
+
+                counts = nplike.zeros(len(nextinputs[0]), dtype=np.int64)
+                all_counts = []
+                all_flatten = []
+                for x in nextinputs:
+
+                    o, f = x._offsets_and_flattened(1, 0)
+                    o = nplike.asarray(o)
+                    c = o[1:] - o[:-1]
+                    nplike.add(counts, c, out=counts)
+                    all_counts.append(c)
+                    all_flatten.append(f)
+
+                offsets = nplike.empty(len(nextinputs[0]) + 1, dtype=np.int64)
+                offsets[0] = 0
+                nplike.cumsum(counts, out=offsets[1:])
+
+                offsets = ak._v2.index.Index64(offsets)
+
+                inner = ak._v2.contents.UnionArray(
+                    ak._v2.index.Index8.empty(len(offsets) - 1, nplike),
+                    ak._v2.index.Index64.empty(len(offsets) - 1, nplike),
+                    all_flatten,
+                )
+
+                tags, index = inner._nested_tags_index(
+                    offsets,
+                    [ak._v2.index.Index64(x) for x in all_counts],
+                )
+
+                inner = ak._v2.contents.UnionArray(tags, index, all_flatten)
+
+                out = ak._v2.contents.ListOffsetArray(
+                    offsets, inner.simplify_uniontype(merge=merge, mergebool=mergebool)
+                )
+
+                return lambda: (out,)
+
+            elif any(
+                x.minmax_depth == (1, 1)
+                for x in inputs
+                if isinstance(x, ak._v2.contents.Content)
+            ):
+                raise ValueError(
+                    "at least one array is not deep enough to concatenate at "
+                    "axis={0}".format(axis)
+                )
+
+            else:
+                return None
+
+        out = ak._v2._util.broadcast_and_apply(
+            contents,
+            getfunction,
+            behavior=ak._v2._util.behavior_of(*arrays, behavior=behavior),
+            allow_records=True,
+            right_broadcast=False,
+            pass_depth=True,
+        )[0]
```

Comment:
This is where the `broadcast_and_apply` from `_broadcasting.py` would go. It didn't have a custom switch for `pass_depth`; that information is one of the many keyword arguments to your `action` function. Use `**kwargs` to absorb all the keyword arguments you don't use (as in `from_regular`).